src/stacked_borrows.rs

```diff
@@ -92,27 +92,90 @@ pub enum Permission {
     Disabled,
 }
 
-/// An item in the per-location borrow stack.
-#[derive(Copy, Clone, Hash, PartialEq, Eq)]
-pub struct Item {
-    /// The permission this item grants.
-    perm: Permission,
-    /// The pointers the permission is granted to.
-    tag: SbTag,
-    /// An optional protector, ensuring the item cannot get popped until `CallId` is over.
-    protector: Option<CallId>,
+impl Permission {
+    fn to_bits(self) -> u64 {
+        match self {
+            Permission::Unique => 0,
+            Permission::SharedReadWrite => 1,
+            Permission::SharedReadOnly => 2,
+            Permission::Disabled => 3,
+        }
+    }
+
+    fn from_bits(perm: u64) -> Self {
+        match perm {
+            0 => Permission::Unique,
+            1 => Permission::SharedReadWrite,
+            2 => Permission::SharedReadOnly,
+            3 => Permission::Disabled,
+            _ => unreachable!(),
+        }
+    }
 }
 
-impl fmt::Debug for Item {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(f, "[{:?} for {:?}", self.perm, self.tag)?;
-        if let Some(call) = self.protector {
-            write!(f, " (call {})", call)?;
+mod item {
+    use super::{Permission, SbTag};
+    use std::fmt;
+
+    /// An item in the per-location borrow stack.
+    #[derive(Copy, Clone, Hash, PartialEq, Eq)]
+    pub struct Item(u64);
+
+    // An Item contains 3 bitfields:
+    // * Bits 0-61 store an SbTag
+    // * Bits 61-63 store a Permission
+    // * Bit 64 stores a flag which indicates if we have a protector
+    const TAG_MASK: u64 = u64::MAX >> 3;
+    const PERM_MASK: u64 = 0x3 << 61;
+    const PROTECTED_MASK: u64 = 0x1 << 63;
+
+    const PERM_SHIFT: u64 = 61;
+    const PROTECTED_SHIFT: u64 = 63;
+
+    impl Item {
+        pub fn new(tag: SbTag, perm: Permission, protected: bool) -> Self {
+            assert!(tag.0.get() <= TAG_MASK);
+            let packed_tag = tag.0.get();
+            let packed_perm = perm.to_bits() << PERM_SHIFT;
+            let packed_protected = (protected as u64) << PROTECTED_SHIFT;
+
+            let new = Self(packed_tag + packed_perm + packed_protected);
+
+            debug_assert!(new.tag() == tag);
```

Comment:
This makes me wonder if we actually ever run this with debug assertions on CI...

Reply:
Ah, that might also be the case. For some time I was super paranoid about integer overflows in particular.

The version shipped via rustup will not have them enabled, though. So I thought this was fine. But for `./miri bench` it's kind of wrong...

Reply:
So more broadly I do agree that it's kind of wrong to have debug assertions on for `./miri bench`. But you should know that this whole function doesn't show up in any profiles I have of Miri.

Reply:
Fair. Anyway that's a separate discussion.